Simplify and speed up table add_column(s)

taldcroft opened the pull request:
This uses the new `_convert_data_to_col` method to simplify and speed up `add_column` so the time for adding is independent of the number of existing table columns.   Details to come, but it is a factor of at least a two or more faster for tables with 20 or more columns.  If you are building a wide table one column at a time the improvement can be a factor of 10 or more.

This flips the previous structure, so that `add_columns` is just calling `add_column` for each new column, instead of `add_column` calling `add_columns` with a one-element list of the new column.  I haven't done timing on `add_columns`, and that might end up being a bit slower.  I think this use-case is less common though.

This implementation allows the data supplied to `add_column(s)` to be any legal data, not just a `Column` or mixin.

This now replaces all the custom `__setitem__` logic so that `t['a'] = data` just calls `t.add_column(data, name='a', copy=True)`.  Before the two paths were essentially independent code-wise, with subtle and not-so-subtle differences.

mhvk commented:
The factoring out generally seems like a good idea - I like `add_column` becoming the more general interface.

taldcroft commented:
## Timing for `add_column()`

This script adds 30 columns (each of length 100) to an empty table.

### Script
```
import numpy as np
import time

from astropy.table import Table, Column
import astropy.table
print(astropy.table.__file__)

c = Column(np.arange(100))

n_col = 30
n_rep = 20
dts = np.zeros((n_rep, n_col))
tots = np.zeros(n_rep)

for i_rep in range(n_rep):
    t = Table()
    times = [time.time()]
    for ii in range(n_col):
        t.add_column(c, name=str(ii))
        times.append(time.time())
    times = np.array(times)
    dts[i_rep] = np.diff(times) * 1000000  # usec
    tots[i_rep] = (times[-1] - times[0]) * 1000000

dts_median = np.median(dts, axis=0)

print('Per column times:')
out = ', '.join(f'{dt:.1f}' for dt in dts_median)
print(out)

print(f'Total: {np.median(tots):.1f}')
```
### This branch
```
(astropy) neptune$ python timing/add_col_timing.py
/Users/aldcroft/git/astropy/astropy/table/__init__.py
Per column times:
77.5, 78.8, 68.5, 68.4, 68.7, 67.4, 66.0, 65.6, 65.6, 65.4, 67.1, 67.5, 69.5, 66.2, 70.0, 69.5, 66.0, 68.9, 69.3, 66.6, 69.4, 68.5, 70.1, 72.0, 66.0, 65.6, 66.5, 65.1, 65.9, 65.6
Total: 2190.6
```
### Release 3.2.1
```
neptune$ python add_col_timing.py
/Users/aldcroft/miniconda3/lib/python3.6/site-packages/astropy/table/__init__.py
Per column times:
113.5, 116.6, 127.0, 140.0, 152.0, 164.0, 175.0, 188.5, 202.1, 214.1, 228.0, 242.8, 255.0, 266.4, 278.0, 291.6, 307.6, 329.3, 334.0, 341.1, 367.4, 384.9, 393.0, 401.9, 415.1, 420.9, 442.1, 447.3, 455.9, 469.0
Total: 8949.0
```

taldcroft commented:
## Timing for `add_columns`

This script uses `add_columns` to add a list of columns, varying the number of original and new columns.

### Script
```
import sys
import numpy as np
import time

from astropy.table import Table, Column

old_col = int(sys.argv[1])
new_col = int(sys.argv[2])
n_rep = 20

dts = []
for i_rep in range(n_rep):
    t = Table()
    for ii in range(old_col):
        t.add_column(Column(np.arange(100)))

    cols = []
    names = []
    for ii in range(new_col):
        cols.append(Column(np.arange(100)))
        names.append(str(ii))
    t0 = time.time()
    t.add_columns(cols, names=names)
    dt = (time.time() - t0) * 1000000  # usec
    dts.append(dt)

dt_median = np.median(dts)
print(f'Median time: {dt_median:.1f} usec')
```
### This branch
```
(astropy) neptune$ python timing/add_cols_timing.py 0 30 # n_original_cols, n_cols added
Median time: 2383.5 usec  # About 10% slower

(astropy) neptune$ python timing/add_cols_timing.py 10 20
Median time: 1609.1 usec  # About the same

(astropy) neptune$ python timing/add_cols_timing.py 20 10
Median time: 808.2 usec  # About 20% faster

(astropy) neptune$ python timing/add_cols_timing.py 28 2
Median time: 180.0 usec  # Factor of 6 faster

(astropy) neptune$ python timing/add_cols_timing.py 2 2
Median time: 159.0 usec  # 20% faster
```
### Release 3.2.1
```
neptune$ python add_cols_timing.py 0 30
Median time: 2162.5 usec

neptune$ python add_cols_timing.py 10 20
Median time: 1595.5 usec

neptune$ python add_cols_timing.py 20 10
Median time: 1050.6 usec

neptune$ python add_cols_timing.py 28 2
Median time: 536.9 usec

neptune$ python add_cols_timing.py 2 2
Median time: 196.0 usec
```

mhvk commented:
Wow!!  It does seem like the above should be added to the benchmarks...

taldcroft commented:
In doing this I discovered that `asdf` is building a Table (via `Table.read`) one column at a time.  This is really the wrong thing to be doing, but I guess the API suggests this is just fine.  

Interestingly, it now looks like adding one column at a time via `setitem` is actually about 5-10% *faster* than a single call to `Table` with a list of cols and names.

mhvk commented on CHANGES.rst:
```diff
@@ -171,6 +183,13 @@ astropy.table
+  methods.  Now it is possible add any object(s) which can be converted or broadcasted
+  to a valid column for the table.  ``Table.__setitem__`` now just calls
+  ``add_column``.  One change is that a value with ``shape = (1,)`` is no longer
+  broadcasted to the column shape. E.g. ``t['a'] = [1]`` used to be the same as
```
Do we really want this? It breaks analogy with an structured array, and it seems somewhat inconsistent with a scalar working. That said, the table is really different in that the first axis is treated as the length. But I don't like special-casing the 1-d case...

taldcroft replied:
OK I can revert this.  Just always seemed to be weird to me, but I don't feel that strongly, and fewer API changes is good.

mhvk commented on astropy/table/table.py:
```diff
@@ -882,10 +882,11 @@ def _convert_data_to_col(self, data, copy=True, default_name=None, dtype=None, n
         # mixin class
-        if (isinstance(data, np.ndarray) and len(data.dtype) > 1 and
-                not self._is_mixin_for_table(data)):
+        if isinstance(data, np.ndarray) and len(data.dtype) > 1 and not is_mixin:
```
Not really for this PR, but I think for most `ndarray` subclasses one could just stick an `info` on and be done with it (can't do this with `ndarray` since it has `__slots__`, not `__dict__`). That might be better than the `.view` which destroys the properties of the subclass.

taldcroft replied:
I don't quite follow.  This is aimed explicitly at structured arrays, which have type of `ndarray`.

mhvk replied:
It is just that the `isinstance` lets through any subclass, which is removed by the `.view`. But I think we can let this be covered by #8944.

mhvk commented on astropy/table/table.py:
```diff
@@ -1711,7 +1667,8 @@ def index_column(self, name):
 
-    def add_column(self, col, index=None, name=None, rename_duplicate=False, copy=True):
+    def add_column(self, col, index=None, name=None, rename_duplicate=False, copy=True,
+                   default_name=None):
```
Should update the docstring to note that `col` does not have to be a `Column` instance.

mhvk commented on astropy/table/table.py:
```diff
@@ -1795,12 +1755,58 @@ def add_column(self, col, index=None, name=None, rename_duplicate=False, copy=Tr
+
+        # If value doesn't have a dtype and won't be added as a mixin then
+        # convert to a numpy array.
+        if not hasattr(col, 'dtype') and not self._is_mixin_for_table(col):
```
Probably again more for follow-up, but in principle it would seem that the property that something should have is `shape` (and perhaps `info`). One could also think of `not isinstance(col, (np.ndarray, ShapedLikeNDArray))`.

(Maybe a follow-up issue on items like this - don't want to hold up this PR, which is mostly a refactor.)

mhvk replied:
For this PR, might it make more sense to have this be part of `_convert_data_to_col`??

taldcroft replied:
Moved to #8943.

taldcroft replied:
> One could also think of not isinstance(col, (np.ndarray, ShapedLikeNDArray))

Currently there is no requirement that a mixin column inherit from `ShapedLikeNDArray`.  That is a very useful convenience, but not required.  (Though in practice I would not be surprised if some things break without that, and I just advised using `transpose()` in the cstack PR).

I thought one idea of duck arrays is to move away from type checking and more toward behavioral consistency.  Of course we are not there yet, but maybe for now this can be left until we change from mixin column to "duck column".  I like that name much better.

mhvk replied:
Really all that's needed is indeed *being* reshapable/broadcastable and being able to deal with info. Anyway, the right place to discuss is the new issue!

taldcroft replied:
> Probably again more for follow-up, but in principle it would seem that the property that something should have is shape (and perhaps info). One could also think of not isinstance(col, (np.ndarray, ShapedLikeNDArray)).

Moved to #8944.

mhvk commented on astropy/table/table.py:
```diff
@@ -1795,12 +1755,58 @@ def add_column(self, col, index=None, name=None, rename_duplicate=False, copy=Tr
+
+        # Make col data shape correct for scalars.  The second test is to allow
+        # broadcasting an N-d element to a column, e.g. t['new'] = [[1, 2]].
+        if ((len(col.shape) == 0 or (len(col.shape) > 1 and col.shape[0] == 1))
```
Now I'm confused again, if we allow `t['new'] = [[1, 2]]` to give a `(n, 2)` shaped column, then why not let `t['new'] = [1]` give a `(n,)` shape column? It thus mean that scalar and shape `(1,)` are treated as the same, but I think that's OK.

mhvk replied:
Overall, I now think I'd go for `if len(self) > 0 and (col.shape == () or col.shape[0] == 1):`

taldcroft replied:
OK will do.  I put the `len(self) > 0` test second because I think that will rarely fail.

mhvk replied:
OK, makes sense.

mhvk commented on astropy/table/table.py:
```diff
@@ -1795,12 +1755,58 @@ def add_column(self, col, index=None, name=None, rename_duplicate=False, copy=Tr
+                self.columns.move_to_end(move_name, last=True)
 
     def add_columns(self, cols, indexes=None, names=None, copy=True, rename_duplicate=False):
         """
```
Also need to update this docstring.

taldcroft commented:
Whew, there was a regression that got caught by a narrative docs test.  I fixed and added a proper test.  With any luck that will be the last test fail.

taldcroft commented:
@bsipocz - should I rebase on master to fix 32 bit or let it slide?

bsipocz commented:
Since there are 96 failing tests that's the same number we saw on master before the workaround, so I would #yolo and ignore it.

taldcroft commented:
Passing tests except expected 32 bit ones.
